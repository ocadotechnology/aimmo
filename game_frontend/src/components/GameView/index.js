import styled from 'styled-components'
import React, { Component } from 'react'
import Unity from 'react-unity-webgl'
import PropTypes from 'prop-types'
import { CircularProgress } from '@material-ui/core'
import { unityContent } from 'api/unity'

export const GameViewLayout = styled.div`
  grid-area: game-view;
`

export const LoadingBackgroundOverlay = styled.div`
  height: 100%;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.2);
`

export const StyledUnity = styled(Unity)`
  display: ${props => props.gameDataLoaded ? 'block' : 'none'};
`

export const StyledCircularProgress = styled(CircularProgress)`
    max-width: 50%;
`

export default class GameView extends Component {
  constructor (props) {
    super(props)
    unityContent.on('loaded', this.unityContentLoaded)
  }

  unityContentLoaded = () => {
    this.props.connectToGame()
  }

  renderLoadingScreen = gameDataLoaded => {
    if (!gameDataLoaded) {
      return (
        <LoadingBackgroundOverlay>
          <StyledCircularProgress color='inherit' />
        </LoadingBackgroundOverlay>
      )
    }
  }

  renderUnityView = (unityContent, gameDataLoaded) => {
    return (
      <StyledUnity
        gameDataLoaded={gameDataLoaded}
        unityContent={unityContent}
        height='100%'
        width='100%' />
    )
  }

  render () {
    return (
      <GameViewLayout>
        {this.renderLoadingScreen(this.props.gameDataLoaded)}
        {this.renderUnityView(unityContent, this.props.gameDataLoaded)}
      </GameViewLayout>
    )
  }
}

GameView.propTypes = {
  connectToGame: PropTypes.func,
  gameDataLoaded: PropTypes.bool
}
