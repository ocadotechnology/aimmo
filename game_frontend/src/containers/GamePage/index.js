import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { actions } from 'features/GhibliMovies'
import Movie from 'components/Movie'
import Button from 'components/Button'

export class GamePage extends Component {
  render () {
    const movieItems = this.props.movies.map(movie =>
      <Movie key={movie.id} movie={movie} />
    )

    return (
      <Fragment>
        <div>Welcome to the AIMMO game screen</div>
        <Button
          id='fetch-movies-button'
          onClick={() => { this.props.fetchMovies() }}>
          Get the GhibliMovies!
        </Button>
        <ul>
          {movieItems}
        </ul>
      </Fragment>
    )
  }
}

GamePage.propTypes = {
  movies: PropTypes.array,
  fetchMovies: PropTypes.func
}

const mapStateToProps = state => {
  return {
    movies: state.movieReducer.movies
  }
}

const mapDispatchToProps = {
  fetchMovies: actions.fetchMovies
}

export default connect(mapStateToProps, mapDispatchToProps)(GamePage)
