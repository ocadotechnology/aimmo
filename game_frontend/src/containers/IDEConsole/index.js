import styled from 'styled-components'
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import LogEntries from 'components/LogEntries'
import ConsoleBar from 'components/ConsoleBar'
import { connect } from 'react-redux'
import { actions } from 'redux/features/ConsoleLog'

export const IDEConsoleSection = styled.section`
  grid-area: ide-console;
  display: flex;
  flex-direction: column;
`

export const StyledConsole = styled.div`
  color: ${props => props.theme.palette.text.primary};
  background-color: ${props => props.theme.palette.background.default};
  font-family: ${props => props.theme.additionalVariables.typography.code.fontFamily};
  overflow: auto;
  white-space: pre-line;
  height: 100%;
  
  ::-webkit-scrollbar {
    background-color: ${props => props.theme.palette.divider};
  }

  ::-webkit-scrollbar-track {
    background: ${props => props.theme.palette.divider}; 
  }

  ::-webkit-scrollbar-thumb {
    background: ${props => props.theme.palette.grey['A200']};
    border-radius: 100px;
    border: ${props => props.theme.spacing(0.25)}px solid transparent;
    background-clip: content-box;
  }
`

export class IDEConsole extends Component {
  static propTypes = {
    logs: PropTypes.arrayOf(PropTypes.object)
  }

  state = {
    scrolledToBottom: true
  }

  handleScroll = event => {
    let target = event.target || event.srcElement
    this.setState({ ...this.state, scrolledToBottom: target.offsetHeight + target.scrollTop === target.scrollHeight })
  }

  componentDidMount () {
    if (this.consoleRef) {
      this.consoleRef.addEventListener('scroll', this.handleScroll)
    }
  }

  componentWillUnmount () {
    this.consoleRef.removeEventListener('scroll', this.handleScroll)
  }

  componentDidUpdate () {
    if (this.lastLogRef && this.state.scrolledToBottom) {
      this.lastLogRef.scrollIntoView(false)
    }
  }
  render () {
    return (
      <IDEConsoleSection>
        <ConsoleBar clearConsoleClicked={this.props.clearConsoleLogs} />
        <StyledConsole innerRef={ref => { this.consoleRef = ref }}>
          <LogEntries
            logs={this.props.logs}
            lastLogRef={ref => { this.lastLogRef = ref }} />
        </StyledConsole>
      </IDEConsoleSection>
    )
  }
}

const mapStateToProps = state => ({
  logs: state.consoleLog.logs
})

const mapDispatchToProps = {
  clearConsoleLogs: actions.clearConsoleLogs
}

export default connect(mapStateToProps, mapDispatchToProps)(IDEConsole)
