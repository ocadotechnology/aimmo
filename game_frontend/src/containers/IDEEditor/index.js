import styled from 'styled-components'
import React, { Component } from 'react'
import AceEditor from 'react-ace'
import 'brace/theme/idle_fingers'
import 'brace/mode/python'
import 'brace/snippets/python'
import 'brace/ext/language_tools'
import { connect } from 'react-redux'
import { actions } from 'features/Editor'
import PropTypes from 'prop-types'

const IDEEditorLayout = styled.div`
  background-color: #2F4F4F
  grid-area: ide-editor
`
export class IDEEditor extends Component {
  render () {
    return (
      <IDEEditorLayout>
        <AceEditor
          mode='python'
          theme='idle_fingers'
          name='ace_editor'
          onLoad={this.props.getCode}
          onChange={this.props.editorChanged}
          fontSize={14}
          showPrintMargin
          showGutter
          highlightActiveLine
          value={this.props.code}
          width='100%'
          height='100%'
          setOptions={{
            enableBasicAutocompletion: true,
            enableLiveAutocompletion: true,
            enableSnippets: true,
            showLineNumbers: true,
            tabSize: 2
          }} />
      </IDEEditorLayout>
    )
  }
}

IDEEditor.propTypes = {
  code: PropTypes.string,
  getCode: PropTypes.func,
  editorChanged: PropTypes.func
}

const mapStateToProps = state => ({
  code: state.editor.code
})

const mapDispatchToProps = {
  getCode: actions.getCodeRequest,
  editorChanged: actions.keyPressed
}

export default connect(mapStateToProps, mapDispatchToProps)(IDEEditor)
