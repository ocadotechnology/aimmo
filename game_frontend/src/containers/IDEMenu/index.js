import styled from 'styled-components'
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { actions } from 'features/Editor'

export const IDEMenuLayout = styled.nav`
  background-color: ${props => props.theme.palette.background.default}
  grid-area: ide-menu
`

export class IDEMenu extends Component {
  render () {
    return (
      <IDEMenuLayout>
        <button
          id='post-code-button'
          onClick={this.props.postCode} >Post Code</button>
      </IDEMenuLayout>
    )
  }
}

IDEMenu.propTypes = {
  postCode: PropTypes.func
}

const mapStateToProps = () => ({})

const mapDispatchToProps = {
  postCode: actions.postCodeRequest
}

export default connect(mapStateToProps, mapDispatchToProps)(IDEMenu)
