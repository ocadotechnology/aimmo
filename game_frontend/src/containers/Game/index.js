import React, { Component, Fragment } from 'react'
import GameView from 'components/GameView'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { actions } from 'redux/features/Game'
import Snackbar from 'components/Snackbar'

export class Game extends Component {
  static propTypes = {
    connectToGame: PropTypes.func,
    theme: PropTypes.object
  }

  state = {
    showSnackbar: this.props.showSnackbarForAvatarUpdated
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.showSnackbarForAvatarUpdated !== this.props.showSnackbarForAvatarUpdated) {
      this.setState({
        showSnackbar: nextProps.showSnackbarForAvatarUpdated
      })
    }
  }

  handleClose = () => {
    this.setState({ showSnackbar: false })
    this.props.snackbarForAvatarUpdatedShown()
  }

  render () {
    return (
      <Fragment>
        <GameView connectToGame={this.props.connectToGame} />
        <Snackbar
          type='success'
          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
          open={this.state.showSnackbar}
          direction='up'
          onClose={this.handleClose}
          message='Your Avatar has been updated with your new code!'
        />
      </Fragment>
    )
  }
}

const mapDispatchToProps = {
  connectToGame: actions.socketConnectToGameRequest,
  snackbarForAvatarUpdatedShown: actions.snackbarForAvatarUpdatedShown
}

const mapStateToProps = state => ({
  showSnackbarForAvatarUpdated: state.game.showSnackbarForAvatarUpdated
})

export default connect(mapStateToProps, mapDispatchToProps)(Game)
