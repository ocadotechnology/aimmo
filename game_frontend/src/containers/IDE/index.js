import React, { Component } from 'react'
import IDEEditor from 'components/IDEEditor'
import IDEConsole from 'components/IDEConsole'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { actions as editorActions } from 'features/Editor'
import { darkTheme } from 'theme'
import { MuiThemeProvider } from '@material-ui/core/styles'
import { ThemeProvider as StyledComponentsThemeProvider } from 'styled-components'

export class IDE extends Component {
  static propTypes = {
    code: PropTypes.string,
    postCode: PropTypes.func,
    getCode: PropTypes.func,
    editorChanged: PropTypes.func,
    logs: PropTypes.arrayOf(PropTypes.object)
  }

  render () {
    return (
      <StyledComponentsThemeProvider theme={darkTheme}>
        <MuiThemeProvider theme={darkTheme}>
          <IDEEditor
            code={this.props.code}
            postCode={this.props.postCode}
            getCode={this.props.getCode}
            editorChanged={this.props.editorChanged}
          />
          <IDEConsole logs={this.props.logs} />
        </MuiThemeProvider>
      </StyledComponentsThemeProvider>
    )
  }
}

const mapStateToProps = state => ({
  code: state.editor.code,
  logs: state.consoleLog.logs
})

const mapDispatchToProps = {
  getCode: editorActions.getCodeRequest,
  editorChanged: editorActions.keyPressed,
  postCode: editorActions.postCodeRequest
}

export default connect(mapStateToProps, mapDispatchToProps)(IDE)
