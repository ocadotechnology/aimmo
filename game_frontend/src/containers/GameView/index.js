import styled from 'styled-components'
import React, { Component } from 'react'
import Unity, { RegisterExternalListener } from 'react-unity-webgl'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { actions } from 'features/Game'

export const GameViewLayout = styled.div`
  grid-area: game-view
`

export class GameView extends Component {
  constructor (props) {
    super(props)

    this.props.getConnectionParameters()

    RegisterExternalListener('SendAllConnect', this.sendAllConnect.bind(this))
  }

  sendAllConnect () {
    this.props.setGameURL(this.props.gameURL)
    this.props.setGamePath(this.props.gamePath)
    this.props.setGamePort(this.props.gamePort)
    this.props.setGameSSL(this.serialisedSSLFlag())
    this.props.establishGameConnection()
  }

  serialisedSSLFlag () {
    let boolString = this.props.gameSSL.toString()

    return boolString.charAt(0).toUpperCase() + boolString.slice(1)
  }

  render () {
    return (
      <GameViewLayout>
        <Unity
          src='/static/unity/Build/unity.json'
          loader='/static/unity/Build/UnityLoader.js'
        />
      </GameViewLayout>
    )
  }
}

GameView.propTypes = {
  gameURL: PropTypes.string,
  gamePath: PropTypes.string,
  gamePort: PropTypes.number,
  gameSSL: PropTypes.bool,
  getConnectionParameters: PropTypes.func,
  setGameURL: PropTypes.func,
  setGamePath: PropTypes.func,
  setGamePort: PropTypes.func,
  setGameSSL: PropTypes.func,
  establishGameConnection: PropTypes.func
}

const mapStateToProps = state => ({
  gameURL: state.game.connectionParameters.game_url_base,
  gamePath: state.game.connectionParameters.game_url_path,
  gamePort: state.game.connectionParameters.game_url_port,
  gameSSL: state.game.connectionParameters.game_ssl_flag
})

const mapDispatchToProps = {
  getConnectionParameters: actions.getConnectionParametersRequest,
  setGameURL: actions.setGameURL,
  setGamePath: actions.setGamePath,
  setGamePort: actions.setGamePort,
  setGameSSL: actions.setGameSSL,
  establishGameConnection: actions.establishGameConnection
}

export default connect(mapStateToProps, mapDispatchToProps)(GameView)
