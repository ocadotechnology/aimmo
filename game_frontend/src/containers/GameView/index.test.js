/* eslint-env jest */
import React from 'react'
import { GameView, GameViewLayout } from 'containers/GameView'
import { shallow } from 'enzyme/build/index'

describe('<GameView />', () => {
  it('matches snapshot', () => {
    const props = {
      gameURL: 'test',
      gamePath: '/test',
      gamePort: 8000,
      gameSSL: false,
      getConnectionParameters: jest.fn(),
      setGameURL: jest.fn(),
      setGamePath: jest.fn(),
      setGamePort: jest.fn(),
      setGameSSL: jest.fn(),
      establishGameConnection: jest.fn()
    }

    const component = shallow(<GameView {...props} />)
    expect(component).toMatchSnapshot()
  })

  it('serialisedSSLFlag function returns correctly', () => {
    const props = {
      getConnectionParameters: jest.fn(),
      gameSSL: false
    }

    const flagReturned = shallow(<GameView {...props} />).instance().serialisedSSLFlag()
    expect(flagReturned).toBe('False')
  })

  it('sendAllConnect function calls all action dispatchers', () => {
    const setGameURL = jest.fn()
    const setGamePath = jest.fn()
    const setGamePort = jest.fn()
    const setGameSSL = jest.fn()
    const establishGameConnection = jest.fn()

    const props = {
      gameURL: 'test',
      gamePath: '/test',
      gamePort: 8000,
      gameSSL: false,
      getConnectionParameters: jest.fn(),
      setGameURL,
      setGamePath,
      setGamePort,
      setGameSSL,
      establishGameConnection
    }

    const wrapper = shallow(<GameView {...props} />)

    wrapper.instance().sendAllConnect()

    expect(setGameURL.mock.calls.length).toBe(1)
    expect(setGamePath.mock.calls.length).toBe(1)
    expect(setGamePort.mock.calls.length).toBe(1)
    expect(setGameSSL.mock.calls.length).toBe(1)
    expect(establishGameConnection.mock.calls.length).toBe(1)
  })
})

describe('<GameViewLayout />', () => {
  it('matches snapshot', () => {
    const tree = shallow(<GameViewLayout />)
    expect(tree).toMatchSnapshot()
  })
})
