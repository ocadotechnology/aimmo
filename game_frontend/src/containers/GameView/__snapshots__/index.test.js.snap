// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameView /> matches snapshot 1`] = `
<styled.div>
  <Unity
    loader="/static/unity/Build/UnityLoader.js"
    src="/static/unity/Build/unity.json"
  />
</styled.div>
`;

exports[`<GameViewLayout /> matches snapshot 1`] = `
.c0 {
  grid-area: game-view;
}

<div
  className="c0"
/>
`;
